cmd/query/app/query_parser.go

```diff
@@ -143,31 +160,125 @@ func (p *queryParser) parse(r *http.Request) (*traceQueryParameters, error) {
+	serviceNames := r.FormValue(servicesParam)
+	if serviceNames == "" {
+		return bqp, newParseError(errors.New("please provide at least one service name"), servicesParam)
+	}
+	bqp.ServiceNames = strings.Split(serviceNames, ",")
+
+	bqp.GroupByOperation, err = p.parseBool(r, groupByOperationParam)
+	if err != nil {
+		return bqp, err
+	}
+	bqp.SpanKinds, err = p.parseSpanKinds(r, spanKindsParam, defaultMetricsSpanKinds)
+	if err != nil {
+		return bqp, err
+	}
+	endTs, err := p.parseTime(r, endTsParam, msDuration)
+	if err != nil {
+		return bqp, err
+	}
+	lookback, err := p.parseDuration(r, lookbackParam, &msDuration, defaultMetricsQueryLookbackDuration)
+	if err != nil {
+		return bqp, err
+	}
+	step, err := p.parseDuration(r, stepParam, &msDuration, defaultMetricsQueryStepDuration)
+	if err != nil {
+		return bqp, err
+	}
+	ratePer, err := p.parseDuration(r, rateParam, &msDuration, defaultMetricsQueryRateDuration)
+	if err != nil {
+		return bqp, err
+	}
+	bqp.EndTime = &endTs
+	bqp.Lookback = &lookback
+	bqp.Step = &step
+	bqp.RatePer = &ratePer
+	return bqp, err
+}
+
+// parseTime parses the time parameter of an HTTP request that is represented the number of "units" since epoch.
+// If the time parameter is empty, the current time will be returned.
+func (p *queryParser) parseTime(r *http.Request, paramName string, units time.Duration) (time.Time, error) {
+	formValue := r.FormValue(paramName)
+	if formValue == "" {
+		if paramName == startTimeParam {
 			return p.timeNow().Add(-1 * p.traceQueryLookbackDuration), nil
 		}
 		return p.timeNow(), nil
 	}
-	micros, err := strconv.ParseInt(value, 10, 64)
+	t, err := strconv.ParseInt(formValue, 10, 64)
 	if err != nil {
-		return time.Time{}, err
+		return time.Time{}, newParseError(err, paramName)
 	}
-	return time.Unix(0, 0).Add(time.Duration(micros) * time.Microsecond), nil
+	return time.Unix(0, 0).Add(time.Duration(t) * units), nil
 }
 
-func (p *queryParser) parseDuration(durationParam string, r *http.Request) (time.Duration, error) {
-	durationInput := r.FormValue(durationParam)
-	if len(durationInput) > 0 {
-		duration, err := time.ParseDuration(durationInput)
-		if err != nil {
-			return 0, fmt.Errorf("cannot not parse %s: %w", durationParam, err)
+// parseDuration parses the duration parameter of an HTTP request that can be represented as either:
+// - a duration string e.g.: "5ms"
+// - a number of units of time e.g.: "1000"
+// If the duration parameter is empty, the given defaultDuration will be returned.
+func (p *queryParser) parseDuration(r *http.Request, paramName string, units *time.Duration, defaultDuration time.Duration) (d time.Duration, err error) {
+	d = defaultDuration
+	formValue := r.FormValue(paramName)
+	switch {
+	case formValue == "":
+		return d, nil
+
+	// If no units are supplied, assume parsing of duration strings like 5ms.
+	case units == nil:
+		if d, err = time.ParseDuration(formValue); err == nil {
```

Comment:
the `search` endpoint defines durations as duration strings, presumably to support human-friendly clients like curl, etc. so this maintains support for such duration parameter values.